core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java

```diff
@@ -354,11 +353,18 @@ private synchronized Map<PlanNodeId, TaskSource> updateSources(List<TaskSource>
         }
 
         // update maxAcknowledgedSplit
-        maxAcknowledgedSplit = sources.stream()
-                .flatMap(source -> source.getSplits().stream())
-                .mapToLong(ScheduledSplit::getSequenceId)
-                .max()
-                .orElse(maxAcknowledgedSplit);
+        for (TaskSource taskSource : sources) {
+            long maxAcknowledgedSplit = taskSource.getSplits().stream()
+                    .mapToLong(ScheduledSplit::getSequenceId)
+                    .max()
+                    .orElse(Long.MIN_VALUE);
+            PlanNodeId planNodeId = taskSource.getPlanNodeId();
+
+            if (!maxAcknowledgedSplitByPlanNode.containsKey(planNodeId)) {
+                maxAcknowledgedSplitByPlanNode.put(planNodeId, Long.MIN_VALUE);
+            }
+            maxAcknowledgedSplitByPlanNode.computeIfPresent(planNodeId, (key, val) -> maxAcknowledgedSplit > val ? maxAcknowledgedSplit : val);
```

Comment:
If taskSource.getSplits().size() == 0 is already checked, it should throw when val >= maxAcknowledgedSplit.

e.g.
```
            maxAcknowledgedSplitByPlanNode.compute(planNodeId, (key, val) -> {
                if (val == null) {
                    return maxAcknowledgedSplit;
                }
                if (val >= maxAcknowledgedSplit) {
                    throw new IllegalStateException(format("%s - splits are out of order? planNodeId=%s, newMax=%d, currentMax=%d", taskId, planNodeId, maxAcknowledgedSplit, val));
                }
                return maxAcknowledgedSplit;
            });
```